homeassistant/components/group.py

```diff
@@ -161,48 +168,78 @@ def reload_service_handler(service_call):
     return True
 
 
-def _process_config(hass, config, component):
+@callback
+def _async_process_config(hass, config, component):
     """Process group configuration."""
+    groups = []
     for object_id, conf in config.get(DOMAIN, {}).items():
         name = conf.get(CONF_NAME, object_id)
         entity_ids = conf.get(CONF_ENTITIES) or []
         icon = conf.get(CONF_ICON)
         view = conf.get(CONF_VIEW)
 
-        group = Group(hass, name, entity_ids, icon=icon, view=view,
-                      object_id=object_id)
-        component.add_entities((group,))
+        group = Group.async_create_group(
+            hass, name, entity_ids, icon=icon, view=view, object_id=object_id)
+        groups.append(group)
 
-    return True
+    hass.loop.create_task(component.async_add_entities(groups))
 
 
 class Group(Entity):
     """Track a group of entity ids."""
 
     # pylint: disable=too-many-instance-attributes, too-many-arguments
-    def __init__(self, hass, name, entity_ids=None, user_defined=True,
-                 icon=None, view=False, object_id=None):
-        """Initialize a group."""
+    def __init__(self, hass, name, user_defined=True, icon=None, view=False):
+        """Initialize a group.
+
+        This Object have factory function for creation.
+        """
         self.hass = hass
         self._name = name
         self._state = STATE_UNKNOWN
-        self._order = len(hass.states.entity_ids(DOMAIN))
         self._user_defined = user_defined
+        self._order = None
         self._icon = icon
         self._view = view
-        self.entity_id = generate_entity_id(
-            ENTITY_ID_FORMAT, object_id or name, hass=hass)
         self.tracking = []
         self.group_on = None
         self.group_off = None
         self._assumed_state = False
-        self._lock = threading.Lock()
-        self._unsub_state_changed = None
+        self._async_unsub_state_changed = None
+
+    @staticmethod
+    # pylint: disable=too-many-arguments
+    def create_group(hass, name, entity_ids=None, user_defined=True,
+                     icon=None, view=False, object_id=None):
+        """Initialize a group."""
+        return run_callback_threadsafe(
+            hass.loop, Group.async_create_group, hass, name, entity_ids,
+            user_defined, icon, view, object_id).result()
+
+    @staticmethod
+    # pylint: disable=too-many-arguments
+    def async_create_group(hass, name, entity_ids=None, user_defined=True,
+                           icon=None, view=False, object_id=None):
+        """Initialize a group.
+
+        This method must be run in the event loop.
+        """
+        self = Group(
+            hass, name, user_defined=user_defined, icon=icon, view=view)
+
+        # init data
+        # pylint: disable=protected-access
```

Comment:
Instead of disabling protected access, we should pass `order` as a parameter into the constructor.